src/explication/explicator/node_explicator.rs

```diff
@@ -299,58 +352,52 @@ fn explicate_encode_do(
             )
         ),
     };
+
+    let node_location = Location::new(value_funclet_id, value_node_id);
     let mut inputs = Vec::new();
-    let mut outputs = Vec::new();
-    for input in expir_inputs.as_ref().opt().expect(&error).iter() {
-        let input_open = input.as_ref().opt().expect(&error).clone();
+    for input in expir_inputs.as_ref().opt().expect(&state.hole_error(context)).iter() {
+        let input_open = input.as_ref().opt().expect(&state.hole_error(context)).clone();
         inputs.push(input_open);
     }
-    for (offset, output) in expir_outputs
-        .as_ref()
-        .opt()
-        .expect(&error)
-        .iter()
-        .enumerate()
-    {
-        let output_open = output.as_ref().opt().expect(&error).clone();
-        let value_location = Location::new(
-            state
-                .get_funclet_spec(
-                    state.get_current_funclet_id(),
-                    &SpecLanguage::Value,
-                    context,
-                )
-                .funclet_id_opt
-                .unwrap(),
-            node_id + offset + 1,
-        );
-        new_state.set_instantiation(
-            output_open,
-            LocationTriple::new_value(value_location),
-            context,
-        );
-        outputs.push(output_open);
-    }
-    let external_function_id = expir_external_funclet_id
-        .as_ref()
-        .opt()
-        .expect(&error)
-        .clone();
-    let node = ir::Node::EncodeDoExternal {
-        encoder,
-        operation,
-        inputs: inputs.into_boxed_slice(),
-        outputs: outputs.into_boxed_slice(),
-        external_function_id,
-    };
-    new_state.next_node();
-    match explicate_node(new_state, context) {
-        None => None,
-        Some(mut out) => {
-            out.add_node(node);
-            Some(out)
+
+    let mut output_attempts = enumerate_output_type_attempts(
```

Comment:
At a quick glance, I'm getting DejaVu looking at this. I think you might have an exact copy-and-paste of this in `local-do-external`? I'm not sure if creating a helper function is the right move here, but at the level of a quick pass it seems like it could be?

Reply:
This is intentional, but I'm impressed you read at this level of detail!  The reason is that the logic inside of these will diverge as I add more "stuff", and making a helper function was already becoming a headache.  I think a helper is ideal, but I'm not 100% sure yet where the logic they share will be, so I've pushed it off

Reply:
That makes sense to me